File: problems/algorithms/2161/C/solution_three_pointers.c


/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *pivotArray(int *nums, int numsSize, int pivot, int *returnSize)
{
	int *result = (int *)malloc(numsSize * sizeof(int));
	if (result == NULL)
		return NULL;
	int lower_count = 0;
	int equal_count = 0;
	for (int i = 0; i < numsSize; i += 1)
	{
		if (nums[i] == pivot)
			equal_count += 1;
		else if (nums[i] < pivot)
			lower_count += 1;
	}
	int lower_index = 0;
	int equal_index = lower_count;
	int higher_index = lower_count + equal_count;
	for (int i = 0; i < numsSize; i += 1)
	{
		if (nums[i] == pivot)
			result[equal_index++] = nums[i];
		else if (nums[i] < pivot)
			result[lower_index++] = nums[i];
		else
			result[higher_index++] = nums[i];
	}
	*returnSize = numsSize;
	return result;
}

File: problems/algorithms/2161/Rust/solution_three_pointers.rs
impl Solution {
    pub fn pivot_array(nums: Vec<i32>, pivot: i32) -> Vec<i32> {
        let mut result = vec![0; nums.len()];
        let mut lower = 0;
        let mut equal = 0;
        for &num in &nums {
            if num == pivot {
                equal += 1;
            } else if num < pivot {
                lower += 1;
            }
        }
        let mut lower_index = 0;
        let mut equal_index = lower;
        let mut higher_index = lower + equal;
        for &num in &nums {
            if num == pivot {
                result[equal_index] = num;
                equal_index += 1;
            } else if num < pivot {
                result[lower_index] = num;
                lower_index += 1;
            } else {
                result[higher_index] = num;
                higher_index += 1;
            }
        }
        result
    }
}